Fix behavior for _index LIKE for ESQL

julian-elastic opened the pull request:
Fixes `_index LIKE <pattern>` to always have normal text matching semantics.

Implement a generic ExpressionQuery and ExpressionQueryBuilder that can be serialized to the data node. Then the  ExpressionQueryBuilder can build an Automaton using TranslationAware.asLuceneQuery() and execute it in Lucine. 

Introduces a breaking change for LIKE on _index fields. The old like behavior is not correct and does not have normal like  semantics from ESQL. Customers upgrading from old build to new build might see a regression, where the data changes due to the like filters on clustering produces different results, but the new results are correct.  

**Behavior for ESQL** 
New CCS to New => New behavior everywhere
Old CCS to New => Old behavior everywhere (the isForESQL flag is not passed in from old)
New CCS to Old => New behavior for new, old behavior for old (the isForESQL cannot be passed, old does not know about it). 
Old CCS to Old => Old behavior everywhere

Closes #129511

julian-elastic committed:
Get basic case to work

julian-elastic committed:
Add ExpressionQueryBuilder with serialization

Commit:
[CI] Auto commit changes from spotless

Commit:
Get _index like working newToNew

julian-elastic committed:
Merge remote-tracking branch 'origin/indexLike_v3' into indexLike_v3

Commit:
[CI] Auto commit changes from spotless

julian-elastic committed:
Move functionality from AutomatonTranslatable to TranslationAware

julian-elastic committed:
Remove AutomatonQueryBuilder and EsqlAutomatonQuery

julian-elastic committed:
Merge remote-tracking branch 'origin/indexLike_v3' into indexLike_v3

julian-elastic committed:
Clean up and comments

julian-elastic committed:
Clean up

julian-elastic committed:
Merge remote-tracking branch 'origin/indexLike_v3' into indexLike_v3

julian-elastic committed:
working on getting the tests to pass in bwc mode

Commit:
[CI] Auto commit changes from spotless

julian-elastic committed:
Merge branch 'main' into indexLike_v3

julian-elastic committed:
Revert "Remove AutomatonQueryBuilder and EsqlAutomatonQuery"

This reverts commit 7484410ea1c880fa72210efd456e14886b5aa131.

julian-elastic committed:
Remove skipped check

julian-elastic committed:
Fix unit tests failing

Commit:
[CI] Auto commit changes from spotless

julian-elastic committed:
Fix transport protocol issues with the unit tests

julian-elastic committed:
Clean up classes no longer needed

Commit:
[CI] Auto commit changes from spotless

julian-elastic committed:
bigfix

julian-elastic committed:
Merge remote-tracking branch 'origin/indexLike_v3' into indexLike_v3

julian-elastic commented on server/src/main/java/org/elasticsearch/action/search/SearchShardsRequest.java:
```diff
@@ -77,7 +78,15 @@ public void writeTo(StreamOutput out) throws IOException {
         out.writeStringArray(indices);
         indicesOptions.writeIndicesOptions(out);
-        out.writeOptionalNamedWriteable(query);
+        out.writeOptionalNamedWriteable(query == null || query.supportsVersion(out.getTransportVersion()) ? query :
```
Do you want to still keep this as part of my PR? I don't think I need it for my changes to work

nik9000 replied:
no, let's drop it.

julian-elastic replied:
Done

julian-elastic commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/Configuration.java:
```diff
@@ -309,4 +313,14 @@ public String toString() {
+     * Please use the constructor that takes a BlockStreamInput instead if you already have a BlockStreamInput
+     */
+    public static Configuration readFrom(StreamInput in) throws IOException {
+        BlockFactory blockFactory = new BlockFactory(new NoopCircuitBreaker(CircuitBreaker.REQUEST), BigArrays.NON_RECYCLING_INSTANCE);
```
Do we need a better solution for this?

julian-elastic committed:
Fix more failing UTs

julian-elastic committed:
Merge branch 'main' into indexLike_v3

julian-elastic committed:
Fix more failing UTs

Commit:
[CI] Auto commit changes from spotless

julian-elastic committed:
Revert change in SearchShardsRequest

julian-elastic committed:
Merge remote-tracking branch 'origin/indexLike_v3' into indexLike_v3

julian-elastic committed:
Revert change in SearchShardsRequest

julian-elastic commented on server/src/main/java/org/elasticsearch/index/query/WildcardQueryBuilder.java:
```diff
@@ -218,7 +233,12 @@ private QueryBuilder maybeRewriteBasedOnConstantFields(@Nullable MappedFieldType
             // since rewrites might happen on a network thread.
-            Query query = constantFieldType.wildcardQuery(value, caseInsensitive, context); // the rewrite method doesn't matter
+            Query query;
+            if (isForESQL) {
```
This is the main change, we call the new constantFieldType.wildcardLikeQuery function for ESQL, otherwise we preserve the old behavior. Notice that I deleted the WildcardLikeQueryBuilder that we created together and opted for a new member in the existing class to simplify the serialization and not worth it to have a whole new class for just one line change.

nik9000 commented:
> In addition, do we want to have a feature flag or other setting to go to the old behavior if a customer complains? This is a breaking change and the customers will see different data.

We can make a cluster setting to turn this on and off. Check out `QUERY_RESULT_TRUNCATION_MAX_SIZE`. The old behavior is so very bug-looking, but we can sure make one of these and check it.

julian-elastic commented on server/src/main/java/org/elasticsearch/index/mapper/ConstantFieldType.java:
```diff
@@ -135,9 +137,46 @@ public final Query wildcardQuery(String value, boolean caseInsensitive, QueryRew
+        SearchExecutionContext context,
+        String description
+    ) {
+        CharacterRunAutomaton compiled = new CharacterRunAutomaton(automaton);
```
I don't think this will improve performance. We call run on the next line. If we get here, we will be calling compiled.run() below in 100% of the cases. Supplier/lazy initialization helps when in some percent of the cases we decide to not materialize the object, and in this case it seems we always need the object. Do you agree? In what circumstances would changing to supplier improve performance?

nik9000 replied:
But of course my comment was wrong. More like:

```
Supplier<CharacterRunAutomaton> runs = new Supplier() {
  CharacterRunAutomaton lazy;

  @Override
  CharacterRunAutomaton get() {
    if (lazy == null) { lazy = new CharacterRunAutomaton(automaton); }
    return lazy;
  }
};
```

julian-elastic replied:
So just to confirm, the supplier itself should be saved in the caller (WildcardLikeList)? One WildcardLikeList object is applied across multiple indices by calling asLuceneQuery/doToQuery multiple times?

julian-elastic replied:
OK, done, also added Supplier for the basic Automation too.

julian-elastic commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/capabilities/TranslationAware.java:
```diff
@@ -47,6 +48,21 @@ static TranslationAware.Translatable translatable(Expression exp, LucenePushdown
+    /**
+     * Translates the implementing expression into a Lucene {@link Automaton}.
+     */
+    default Automaton asLuceneQuery() {
```
Done

julian-elastic commented on server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java:
```diff
@@ -370,6 +384,22 @@ public Query regexpQuery(
+     * Returns a Lucine pushable Query for the current field
+     * For now can only be AutomatonQuery or MatchAllDocsQuery() or MatchNoDocsQuery()
+     */
+    public Query automatonQuery(
```
True, it is AutomatonQuery for keyword. What I meant is that there are only those 3 Query types returned in any of the sub classes.
How do you want to rephrase this?

nik9000 replied:
ah, I see. it's good. I was just reading it wrong.

julian-elastic committed:
Address code review feedback part 1

julian-elastic committed:
Merge branch 'main' into indexLike_v3

julian-elastic committed:
Update docs/changelog/130019.yaml

julian-elastic committed:
Address code review feedback part 2 (still needs UT with the flag set to false)

julian-elastic committed:
Merge remote-tracking branch 'origin/indexLike_v3' into indexLike_v3

Commit:
[CI] Auto commit changes from spotless

julian-elastic committed:
UT fixes

julian-elastic committed:
Merge remote-tracking branch 'origin/indexLike_v3' into indexLike_v3

julian-elastic committed:
Merge branch 'main' into indexLike_v3

julian-elastic committed:
Fix merge errors

julian-elastic committed:
Add Supplier for Automaton and CharacterRunAutomaton

julian-elastic committed:
Fix UT errors

julian-elastic committed:
Fix UT errors

julian-elastic committed:
Change behavior for LIKE LIST and _index

julian-elastic commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/regex/WildcardLikeList.java:
```diff
@@ -106,22 +150,41 @@ public Translatable translatable(LucenePushdownPredicates pushdownPredicates) {
      */
     @Override
     public Query asQuery(LucenePushdownPredicates pushdownPredicates, TranslatorHandler handler) {
+        if (configuration != null && configuration.stringLikeOnIndex() == false) {
```
It seems we don't really have a good option here. After discussion with Nik, we decided to keep the setting for now.  It will be hidden and we will only provide it to customers who request the old LIKE behavior.

nik9000 committed:
Plan serialization?

nik9000 committed:
fixup

nik9000 committed:
Merge remote-tracking branch 'julian-elastic/indexLike_v3' into indexLike_v3

julian-elastic committed:
Address no commits

julian-elastic committed:
Fix failing UTs

Commit:
[CI] Auto commit changes from spotless

julian-elastic committed:
Merge branch 'main' into indexLike_v3

julian-elastic committed:
Remove some debugging print statements

julian-elastic commented:
I have addressed the 3 NOCOMMITS. 
ExpressionBuilder.java - Removed the configuration, it is no longer needed. 

Configuration.java - Tried to move the setting from the Configuration to  LucenePushDownPredicate
Got it working for LucenePushDownPredicate.from()
Did not get it working for LucenePushDownPredicate.forCanMatch(). I think that is where it is actually. How do I get the setting value from DataNodeComputeHandler::startComputeOnDataNodes()? 

FilterTests.java
Added the EsqlTestUtils.TEST_CFG so it is no longer null. 
Also there was a bug with a newly added test, I fixed it by using an older version so serialization is not available. Please review the fix.

Commit:
[CI] Auto commit changes from spotless

nik9000 committed:
Move flag to EsqlFlags

nik9000 committed:
Merge remote-tracking branch 'julian-elastic/indexLike_v3' into indexLike_v3

julian-elastic committed:
Clean up and fix UTs

Commit:
[CI] Auto commit changes from spotless

julian-elastic committed:
Fix UT fail

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/PlannerUtils.java:
```diff
@@ -253,7 +261,13 @@ && translatable(exp, ctx).finish() == TranslationAware.FinishedTranslatable.YES)
+                    QueryBuilder builder = qlQuery.toQueryBuilder();
+                    LogManager.getLogger(PlannerUtils.class).error("ASFAFDSAFDSF {} {}", qlQuery, qlQuery.containsPlan());
+                    if (qlQuery.containsPlan()) {
+                        builder = new PlanStreamWrapperQueryBuilder(configuration, builder);
```
> e.g. sometimes we have containsPlan() returning next.containsPlan().

Yes.

>It should be fine

Your second question - do we have an integration test for it - we do, all the ones you wrote. But should we have something else? I don't... think so? But if you have any ideas I'm happy to brainstorm.

elasticsearchmachine commented:
Hi @julian-elastic, I've updated the changelog YAML for you.

julian-elastic committed:
Update docs/changelog/130019.yaml

julian-elastic committed:
Address code review comments

julian-elastic commented on server/src/main/java/org/elasticsearch/index/query/WildcardQueryBuilder.java:
```diff
@@ -56,6 +56,9 @@ public class WildcardQueryBuilder extends AbstractQueryBuilder<WildcardQueryBuil
     private static final ParseField CASE_INSENSITIVE_FIELD = new ParseField("case_insensitive");
     private boolean caseInsensitive = DEFAULT_CASE_INSENSITIVITY;
 
+    // forces a string like match instead of a wildcard match
```
Done

julian-elastic commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/ExpressionQueryBuilder.java:
```diff
@@ -0,0 +1,123 @@
+
+/**
+ * Implements an Expression query builder, which matches documents based on a given expression.
+ * The expression itself must provide the asLuceneQuery and getLuceneQueryDescription methods
```
Done

julian-elastic commented:
Merged with https://github.com/elastic/elasticsearch/pull/130849
